macros/src/ast_node.rs

```diff
@@ -0,0 +1,172 @@
+                new_def.to_tokens(stream);
+            }
+        }
+    }
+
+    pub fn generate_transitive_from_impls(&self, stream: &mut TokenStream) {
+        let root = &self.enums[0].ident;
+
+        for category in &self.tree[root] {
+            for node in &self.tree[category] {
+                if let Some(leaves) = self.tree.get(node) {
+                    // From<KeywordEnum> for NodeEnum
+                    self.emit_from_impl([node, category, root], true, stream);
+
+                    for leaf in leaves {
+                        // From<KeywordTime> for LeafEnum
+                        self.emit_from_impl([leaf, node, category], false, stream);
+                        // From<KeywordTime> for NodeEnum
+                        self.emit_from_impl([leaf, category, root], false, stream);
+                    }
+                } else {
+                    // From<Argument> for NodeEnum>
+                    self.emit_from_impl([node, category, root], false, stream);
+                }
+            }
+        }
+    }
+
+    fn emit_from_impl(
+        &self,
+        [inner, middle, outer]: [&Ident; 3],
+        inner_is_already_enum: bool,
+        stream: &mut TokenStream,
+    ) {
+        for flavor in FLAVORS {
+            let inner_ty = if inner_is_already_enum {
+                flavor.suffix_and_lifetime(inner)
+            } else {
+                flavor.ampersand(inner)
+            };
+            let middle_ty = flavor.suffix(middle);
+            let outer_ty = flavor.suffix_and_lifetime(outer);
+
+            let lifetime = (flavor != Flavor::Val).then_some(quote!(<'n>));
+
+            quote! {
+                impl #lifetime From<#inner_ty> for #outer_ty {
+                    fn from(value: #inner_ty) -> Self {
+                        Self::from(#middle_ty::from(value))
+                    }
+                }
+            }
+            .to_tokens(stream)
+        }
+    }
+
+    /// Hackish: An accept() method that takes NodeRef<'a> rather than &'a NodeRef<'_>.
+    /// Visitors that actually use the lifetime, such as Traversal, need this behavior,
+    /// because if you have an &'a NodeRef<'b>, then 'b is your actual borrow from the AST.
+    pub fn generate_inherent_accept_method(&self, stream: &mut TokenStream) {
+        for definition in self.enums {
+            let ty = Flavor::Ref.suffix(&definition.ident);
+            let variants = definition.variants.iter().map(|v| &v.ident);
+            quote! {
+                impl<'n> #ty<'n> {
+                    pub fn accept(self, visitor: &mut impl NodeVisitor<'n>, reverse: bool) {
+                        match self {
+                            #(Self::#variants(x) => x.accept(visitor, reverse)),*
+                        }
+                    }
+                }
+            }
+            .to_tokens(stream)
+        }
+    }
+}
+
+#[derive(Copy, Clone, PartialEq)]
+enum Flavor {
+    Val,
+    Ref,
+    Mut,
```

Comment:
isn't this usually called RefMut? Maybe use `Value` instead of `Val` then.

Reply:
This enum's days are numbered anyway: we only use `NodeEnum` for the AST root, and we only use `NodeRefMut` for parsing. In both cases, it's private to the `ast` module and I've already described a refactor that would eliminate the enum.